Generalize tonatona db

cdepillabout opened the pull request:
This PR generalizes tonatona-db so that it can work with multiple DBs.

This PR adds support for both PostgreSQL and SQLite.

arowM commented on tonatona-sample/src/Tonatona/Sample.hs:
```diff
@@ -121,16 +137,21 @@ instance TonaServantConfig Config where
 
 data Shared = Shared
-  { tonaDb :: TonaDb.Shared
+  { tonaDb :: TonaDb.SharedSql
```
Architecture として命名に規則性をもたせたい(`Shared` っていう名前で統一したい)ので、
`import qualified Tonatona.Db.Sql as TonaDbSql` で import して
`tonaDbSql :: TonaDb.Shared`
みたいな形にすることは可能でしょうか？

cdepillabout replied:
You're right, I'll change this!

arowM commented:
It's amazing to making cool module structure.
Please check my trivial comment about regularity.

cdepillabout commented:
@arowM 

> It is recommended to keep regularity for making better architecture, so please use the word Shared as bellow.
> 
> ```haskell
> import qualified Tonatona.Db.Sql as TonaDbSql
> ...
> 
> data Shared = Shared
>   { tonaDbSql :: TonaDb.Shared
>   ...
> ```

I fixed this in commit 325eefd83.

However, I remembered why I original used the name `SharedSql` instead of just `Shared`.

`Tonatona.Db` originally creates the `Shared` type:

```haskell
data Shared backend = Shared
  { runDb :: ...
  }
```

Then in `Tonatona.Db.Sql` I create a type synonym for an SQL backend:

```haskell
type SharedSql = TonaDb.Shared SqlBackend
```

Ideally, I would want to call this type synonym `Shared` like the following:

```haskell
import qualified Tonatona.Db as TonaDb

type Shared = TonaDb.Shared SqlBackend
```

The problem is that I want to export the *constructor* for `TonaDb.Shared`, but not the datatype `TonaDb.Shared`.  Instead of exporting the dataype, I would just export the type synonym `Shared`.  In `Tonatona.Db.Sql`:

```haskell
module Tonatona.Db.Sql
  ( TonaDb.Shared(Shared)
  , Shared
  , ...
  ) where
```

However, this is not legal Haskell.  Haskell doesn't let you export *both* a type synonym and a data type that are the same name.  There is also no way to export *only* a data constructor.

I fixed this by adding a smart constructor for `Shared` called `mkShared`.  The `Tonatona.Db.Sql.Sqlite` and `Tonatona.Db.Sql.PostgreSQL` modules also export an`init` function that lets you easily create a `Shared` data type.

cdepillabout commented:
This should be good to merge in.

arowM commented:
I got it.
Great! Thanks.
